src/Analyser/TypeSpecifier.php

```diff
@@ -680,14 +681,100 @@ public function specifyTypesInCondition(
 					throw new ShouldNotHappenException();
 				}
 
-				return array_reduce(
-					array_filter(
-						$expr->vars,
-						static fn (Expr $var) => $scope->issetCheck($var, static fn () => true),
-					),
-					fn (SpecifiedTypes $types, Expr $var) => $types->unionWith($this->specifyTypesInCondition($scope, $var, $context, $rootExpr)),
-					new SpecifiedTypes(),
-				);
+				$specifiedTypes = new SpecifiedTypes();
+				foreach ($expr->vars as $var) {
+					$isset = $scope->issetCheck($var, static fn () => true);
+
+					// variable is always defined
+					if ($isset === true) {
+						$specifiedTypes = $specifiedTypes->unionWith($this->specifyTypesInCondition($scope, $var, $context, $rootExpr));
+						continue;
+					}
+
+					if (!($var instanceof ArrayDimFetch)
+						|| $var->dim === null
+					) {
+						continue;
+					}
+
+					// variable is always undefined
+					if (
+						$isset === false
+					) {
+						continue;
+					}
+
+					// variable maybe be defined
+					$type = $scope->getType($var->var);
+					if ($type instanceof MixedType) {
+						continue;
+					}
+
+					$dimType = $scope->getType($var->dim);
+					if (!($dimType instanceof ConstantIntegerType) && !($dimType instanceof ConstantStringType)) {
+						continue;
+					}
+
+					$hasOffsetType = $type->hasOffsetValueType($dimType);
+					$offsetType = $type->getOffsetValueType($dimType);
+					if ($hasOffsetType->yes() && !TypeCombinator::containsNull($offsetType)) {
+						$specifiedTypes = $specifiedTypes->unionWith($this->create(
+							new NotIssetExpr($var),
+							new ErrorType(),
+							$context,
+							false,
+							$scope,
+							$rootExpr,
+						));
+						continue;
+					}
+
+					if (TypeCombinator::containsNull($offsetType)) {
+						$newType = $type->unsetOffset($dimType);
+
+						$newType = TypeTraverser::map($newType, static function (Type $type, callable $traverse) use ($dimType): Type {
+							if ($type instanceof UnionType || $type instanceof IntersectionType) {
+								return $traverse($type);
+							}
+
+							if ($type instanceof ConstantArrayType) {
+								$builder = ConstantArrayTypeBuilder::createFromConstantArray(
+									$type,
+								);
+								$builder->setOffsetValueType(
+									$dimType,
+									new NullType(),
+									true,
+								);
+								return $builder->getArray();
+							}
+
+							return $type->setOffsetValueType($dimType, new NullType());
+						});
+
+						$specifiedTypes = $specifiedTypes->unionWith($this->create(
+							$var->var,
+							$newType,
+							$context->negate(),
+							true,
```

Comment:
I think there is no way arround `override=true` here, because we do not intersect/union the previous known array-shape, but just update a single offset of it.